### PR TITLE
fix(version): Version::init() race fix

### PR DESCRIPTION
**Pull request by ruck314**
## Summary

Guards `Version::init()` with `std::call_once` and declares its statics atomic so concurrent first-callers cannot observe partially-initialised state.

## Bugs fixed

- fix(CORE-005): guard Version::init() with call_once and declare statics atomic

## Verification

- Lint: `./scripts/run_linters.sh`
- Python tests: `pytest tests/`

## Notes

The previously included C++ audit-style repro test was removed per maintainer feedback - that style of test was deemed too brittle relative to the value it provides. Functional coverage for this fix is difficult given its low-level nature; the audit/guardrail intent has been folded into AI agent review directives instead.

### CODE REVIEW

File: src/rogue/Version.cpp [hunk 1]
```diff
@@ -34,28 +35,36 @@
...
+        char lead;
+        uint32_t maj = 0, min = 0, mnt = 0, dev = 0;
+        int32_t ret  = sscanf(_version,
+                              "%c%" SCNu32 ".%" SCNu32 ".%" SCNu32 "-%" SCNu32 "-%s",
+                              &lead,
+                              &maj,
+                              &min,
+                              &mnt,
+                              &dev,
+                              dump);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/Version.cpp:58]: Fixed — changed `%s` to `%99s` to bound the write into `dump[100]`.

File: include/rogue/Version.h [hunk 2]
```diff
@@ -20,6 +20,8 @@
 
 #include <stdint.h>
 
+#include <atomic>
+#include <mutex>
```
[the remark this answers is not in the document]
Comment (ruck314) [reply include/rogue/Version.h]: Fixed — removed `<mutex>` from the header; it's only needed in `Version.cpp`.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#include "doctest/doctest.h"
+#include "rogue/Version.h"
+
+namespace {
+
+std::string readFile(const std::string& path) {
+    std::ifstream f(path);
+    if (!f.is_open()) return "";
+    return std::string((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());
+}
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Fixed — added explicit `#include <iterator>` for portability.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ * init() before reading these statics. Two concurrent callers can race,
+ * producing torn reads or torn writes on platforms where multi-word stores
+ * are not atomic.
+ *
+ * Because the race is undefined behaviour and requires Thread Sanitizer or
+ * precise ordering to detect at runtime, this test uses a 100% deterministic
+ * source-text invariant: reads Version.cpp and asserts that `_major` is
+ * declared std::atomic<uint32_t> or that init() is guarded by
+ * std::call_once / a mutex. On HEAD neither exists → FAIL fires.
+ *
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Fixed — rewrote the description to frame this as a regression guard for the historical race, not a description of current (unfixed) state.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+TEST_CASE("Version::init() protects static member writes from concurrent access") {
+    // Read Version.cpp and Version.h and verify that the static _major/_minor/_maint/_devel
+    // are declared std::atomic or that init() is called at most once via std::call_once.
+    //
+    // On HEAD:
+    //   - Version.cpp declares `uint32_t rogue::Version::_major = 0;` (plain uint32_t)
+    //   - init() unconditionally re-assigns _major/_minor/_maint/_devel
+    //   - No std::call_once or mutex is used in init()
+    // → FAIL_CHECK fires deterministically.
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Fixed — replaced the stale "On HEAD" block with a concise regression-guard description referencing CORE-005, and corrected the macro name to match the actual CHECK_MESSAGE usage.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    // or as a private static std::atomic member.
+    bool major_is_atomic = (src_h.find("atomic<uint32_t>") != std::string::npos &&
+                            src_h.find("_major")           != std::string::npos)
+                        || (src_h.find("std::atomic")      != std::string::npos &&
+                            src_h.find("_major")           != std::string::npos);
+
+    CHECK_MESSAGE(major_is_atomic,
+                  "Version::_major is not declared std::atomic<uint32_t>; "
+                  "concurrent calls to init() introduce a data race on _major/_minor/_maint/_devel. "
+                  "Declare the statics as std::atomic<uint32_t> or guard init() with std::call_once.");
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Fixed — changed "or" to "and" since the test independently verifies both the atomic declarations and the call_once guard.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    // Check 2: init() should be idempotent — guarded by std::call_once or an initialised-flag check.
+    bool has_call_once  = src_cpp.find("call_once")   != std::string::npos;
+    bool has_once_flag  = src_cpp.find("once_flag")   != std::string::npos;
+    bool has_init_guard = has_call_once || has_once_flag;
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Disagree — this is a pragmatic regression guard on a small, well-known file (Version.cpp is ~80 lines). The only way "call_once" or "once_flag" would appear in a comment is if someone was documenting the guard itself, which is a safe match. Stripping comments would require a C++ comment parser (or regex heuristics with their own edge cases), adding fragility for zero practical benefit. The false-positive risk is negligible for this use case.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    bool major_is_atomic = (src_h.find("atomic<uint32_t>") != std::string::npos &&
+                            src_h.find("_major")           != std::string::npos)
+                        || (src_h.find("std::atomic")      != std::string::npos &&
+                            src_h.find("_major")           != std::string::npos);
+
+    CHECK_MESSAGE(major_is_atomic,
+                  "Version::_major is not declared std::atomic<uint32_t>; "
+                  "concurrent calls to init() introduce a data race on _major/_minor/_maint/_devel. "
+                  "Declare the statics as std::atomic<uint32_t> and guard init() with std::call_once.");
+
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Fixed — extended the audit to independently verify all four members (_major, _minor, _maint, _devel) are declared std::atomic<uint32_t>.

File: tests/cpp/core/CMakeLists.txt [hunk 9]
```diff
@@ -21,3 +21,13 @@ rogue_add_cpp_test(rogue-cpp-ctor-catch-thread-cleanup
...
+
+rogue_add_cpp_test(rogue-cpp-version-init-race-audit-repro
+   SOURCES
+      test_version_init_race_audit_repro.cpp
+   LABELS
+      cpp-core
+      no-python
+)
+target_compile_definitions(rogue-cpp-version-init-race-audit-repro
+   PRIVATE ROGUE_SRC_DIR="${CMAKE_SOURCE_DIR}")
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/CMakeLists.txt]: Fixed — changed to PROJECT_SOURCE_DIR so the path resolves correctly when rogue is consumed via add_subdirectory or FetchContent.

File: tests/cpp/core/test_version_init_race_audit_repro.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    REQUIRE_MESSAGE(!src_h.empty(),
+                    "could not open include/rogue/Version.h (ROGUE_SRC_DIR=",
+                    ROGUE_SRC_DIR, ")");
+
+    // Check 1: all four version statics must be declared std::atomic<uint32_t>.
+    auto is_atomic_member = [&](const char* name) {
+        return (src_h.find("atomic<uint32_t>") != std::string::npos &&
+                src_h.find(name)               != std::string::npos)
+            || (src_h.find("std::atomic")      != std::string::npos &&
+                src_h.find(name)               != std::string::npos);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/cpp/core/test_version_init_race_audit_repro.cpp]: Fixed — now searches for the combined pattern `atomic<uint32_t> _name` as a single substring, so reverting any individual member to plain `uint32_t` will correctly fail the audit.

**Review by bengineerd — APPROVED** [review]
